examples/temporal_graph_learning/gnn_encoder.py

```diff
@@ -0,0 +1,218 @@
+
+    Parameters
+    ----------
+    g: DistGraph
+        The distributed graph
+    in_feat : int
+        Input dimension
+    out_dim : int
+        Output dimension
+    num_heads : int
+        Number of attention heads
+    activation: torch.nn.Module
+        Activation function.
+    dropout : float
+        Dropout. Default 0.
+    """
+    def __init__(self, g, in_feat, out_feat, num_heads, activation=None, dropout=0.0):
+        super(ResidualTemporalRelationalGraphConv, self).__init__()
+        self.in_feat = in_feat
+        self.out_feat = out_feat
+        self.activation = activation
+        self.dropout = nn.Dropout(dropout)
+
+        self.conv = TemporalRelationalGraphConv(g, in_feat, out_feat, num_heads)
+        self.loop_weights = nn.Linear(in_feat, out_feat, bias=True)
+
+    def forward(self, g, inputs):
+        """Forward computation. Residual connection is used.
+
+        Parameters
+        ----------
+        g: DGL MFGs
+            Sampled subgraph in DGL MFG
+        inputs: dict[str, torch.Tensor]
+            Input node feature for each node type.
+        """
+        g = g.local_var()
+        hs = self.conv(g, inputs)
+
+        # Add residual connection, activation, dropout
+        outputs = {}
+        for ntype, ntype_inputs in inputs.items():
+            outputs[ntype] = {}
+            for nfield, ntype_nfield_inputs in ntype_inputs.items():
+                # get the residual feats
+                h = self.loop_weights(
+                    ntype_nfield_inputs[: g.number_of_dst_nodes(ntype)]
+                )
+                if ntype in hs and nfield in hs[ntype]:
+                    h = h + hs[ntype][nfield]
+                if self.activation != None:
+                    h = self.activation(h)
+                outputs[ntype][nfield] = self.dropout(h)
+
+        return outputs
+
+class TemporalRelationalGraphConv(nn.Module):
+    """ Relational temporal graph conv layer without residual connection.
+
+    Parameters
+    ----------
+    g: DistGraph
+        The distributed graph
+    in_feat : int
+        Input dimension
+    out_dim : int
+        Output dimension
+    num_heads : int
+        Number of attention heads
+    """
+    def __init__(self, g, in_feat, out_feat, num_heads):
+        super(TemporalRelationalGraphConv, self).__init__()
+        self.in_feat = in_feat
+        self.out_feat = out_feat
+        self.node_fields = get_unique_nfields(g.etypes)
+
+        self.conv = nn.ModuleDict()
+        for canonical_etypes in get_unique_etype_triplet(g.canonical_etypes):
+            conv_name = "%s-%s-%s" % (
+                canonical_etypes[0],
+                canonical_etypes[1],
+                canonical_etypes[2],
+            )
+            gc_layer = GATConv(
+                in_feat,
+                out_feat // num_heads,
+                num_heads,
+                bias=False,
+                allow_zero_in_degree=True,
+            )
```

Comment:
this implementation also uses the model parameters to distinguish different timestamps.

Reply:
No. `canonical_etypes` is returned by `get_unique_etype_triplet`, which removed the timestamp information on the edge type.

In this example, we have
```
self.conv = {
       'paper-cite-paper': dgl.nn.GATConv
}
```